salt/utils/network.py

```diff
@@ -1906,3 +1906,59 @@ def dns_check(addr, port, safe=False, ipv6=None):
             raise SaltClientError()
         raise SaltSystemExit(code=42, msg=err)
     return resolved
+
+
+def parse_host_port(host_port):
```

Comment:
Uhm... But do we really need all that? You also didn't wrote unit tests for it and yet you have a bug inside of the code. I don't really in favour of all this, to be honest. IPv6 ports are separated with `[` and `]`. So you don't need to maintain all these monstrous regular expressions, but just snip off the port away and let `ipaddress` does the rest. Something like this (untested, I wrote it directly here):

```python
from salt._compat import ipaddress

...

def get_host_port(hostport):
    host = port = None
    if not any([hostport.endswith(p) for p in ['.', ':']]):
        if ']:' in hostport:
            host, port = hostport[1:].split(']:')
        elif ':' in hostport and '.' in hostport:
            host, port = hostport.split(':')
        try:
            port = int(port)
        except ValueError, err:
            log.debug('Not a port: %s', err)
        try:
            host = ipaddress.ip_address(host)
        except ValueError, err:
            log.debug(err)

    return host, port
```
Look, ma! No regexes!

The rest `ipaddress` should just do. If it raises and exception "not an IPv6 or IPv4 address", then it is a host. Otherewise you will get an object instance where `.version` will tell you what it is.